cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -309,6 +314,208 @@ Result<std::vector<FileInfo>> LocalFileSystem::GetFileInfo(const FileSelector& s
+          }
+        }
+        return status;
+      }
+      child_fns_ = result.MoveValueUnsafe();
+
+      const size_t dirent_count = child_fns_.size();
+      current_chunk_.reserve(dirent_count >= batch_size_ ? batch_size_ : dirent_count);
+
+      initialized_ = true;
+      return Status::OK();
+    }
+
+    Result<FileInfoVector> Next() {
+      if (!initialized_) {
+        auto init = Initialize();
+        if (!init.ok()) {
+          return Finish(init);
+        }
+      }
+      while (idx_ < child_fns_.size()) {
+        auto full_fn = dir_fn_.Join(child_fns_[idx_++]);
+        auto res = StatFile(full_fn.ToNative());
+        if (!res.ok()) {
+          return Finish(res.status());
+        }
+
+        auto info = res.MoveValueUnsafe();
+
+        // Try to recurse into subdirectories, if needed.
+        if (info.type() == FileType::Directory &&
+            nesting_depth_ < selector_.max_recursion && selector_.recursive) {
+          auto status = DoDiscovery(std::move(full_fn), nesting_depth_ + 1, selector_,
+                                    file_gen_producer_, batch_size_);
+          if (!status.ok()) {
+            return Finish(status);
+          }
+        }
+        // Everything is ok. Add the item to the current chunk of data.
+        current_chunk_.emplace_back(std::move(info));
+        // Keep `current_chunk_` as large, as `batch_size_`.
+        // Otherwise, yield the complete chunk to the caller.
+        if (current_chunk_.size() == batch_size_) {
+          FileInfoVector yield_vec;
+          std::swap(yield_vec, current_chunk_);
+          const size_t items_left = child_fns_.size() - idx_;
+          current_chunk_.reserve(items_left >= batch_size_ ? batch_size_ : items_left);
+          return yield_vec;
+        }
+      }  // while (idx_ < child_fns_.size())
+
+      // Flush out remaining items
+      if (!current_chunk_.empty()) {
+        FileInfoVector yield_vec;
+        std::swap(yield_vec, current_chunk_);
+        return yield_vec;
+      }
+      return Finish();
+    }
+
+   private:
+    /// Close the producer end of stream and return iteration end marker.
+    Result<FileInfoVector> Finish(Status status = Status::OK()) {
+      file_gen_producer_.Close();
+      ARROW_RETURN_NOT_OK(status);
+      return IterationEnd<FileInfoVector>();
+    }
+  };
+
+  /// Create an instance of  `DiscoveryImplIterator` under the hood for the
+  /// specified directory, wrap it in the `BackgroundGenerator + TransferredGenerator`
+  /// bundle and feed the results to the main producer queue.
+  static Status DoDiscovery(const PlatformFilename& dir_fn, int32_t nesting_depth,
+                            FileSelector selector,
+                            FileInfoGeneratorProducer file_gen_producer,
+                            uint32_t batch_size) {
+    ARROW_RETURN_IF(file_gen_producer.is_closed(),
+                    arrow::Status::Cancelled("Discovery cancelled"));
```

Comment:
When does this occur?

Reply:
Well, this is more of a general failsafe rather than motivated by some real possibility. Ideally, should not happen, since there's no way to interact with the producer directly from the user's point of view. But, it's always better be safe than sorry. :)